src/EventIndexing.js

```diff
@@ -0,0 +1,406 @@
+
+            if (cancelled) {
+                break;
+            }
+
+            const checkpoint = this.crawlerCheckpoints.shift();
+
+            /// There is no checkpoint available currently, one may appear if
+            // a sync with limited room timelines happens, so go back to sleep.
+            if (checkpoint === undefined) {
+                continue;
+            }
+
+            console.log("EventIndex: crawling using checkpoint", checkpoint);
+
+            // We have a checkpoint, let us fetch some messages, again, very
+            // conservatively to not bother our Homeserver too much.
+            const eventMapper = client.getEventMapper();
+            // TODO we need to ensure to use member lazy loading with this
+            // request so we get the correct profiles.
+            let res;
+
+            try {
+                res = await client._createMessagesRequest(
+                    checkpoint.roomId, checkpoint.token, this._eventsPerCrawl,
+                    checkpoint.direction);
+            } catch (e) {
+                console.log("EventIndex: Error crawling events:", e);
+                this.crawlerCheckpoints.push(checkpoint);
+                continue;
+            }
+
+            if (res.chunk.length === 0) {
+                console.log("EventIndex: Done with the checkpoint", checkpoint);
+                // We got to the start/end of our timeline, lets just
+                // delete our checkpoint and go back to sleep.
+                await indexManager.removeCrawlerCheckpoint(checkpoint);
+                continue;
+            }
+
+            // Convert the plain JSON events into Matrix events so they get
+            // decrypted if necessary.
+            const matrixEvents = res.chunk.map(eventMapper);
+            let stateEvents = [];
+            if (res.state !== undefined) {
+                stateEvents = res.state.map(eventMapper);
+            }
+
+            const profiles = {};
+
+            stateEvents.forEach(ev => {
+                if (ev.event.content &&
+                    ev.event.content.membership === "join") {
+                    profiles[ev.event.sender] = {
+                        displayname: ev.event.content.displayname,
+                        avatar_url: ev.event.content.avatar_url,
+                    };
+                }
+            });
+
+            const decryptionPromises = [];
+
+            matrixEvents.forEach(ev => {
+                if (ev.isBeingDecrypted() || ev.isDecryptionFailure()) {
+                    // TODO the decryption promise is a private property, this
+                    // should either be made public or we should convert the
+                    // event that gets fired when decryption is done into a
+                    // promise using the once event emitter method:
+                    // https://nodejs.org/api/events.html#events_events_once_emitter_name
+                    decryptionPromises.push(ev._decryptionPromise);
```

Comment:
We tend to go with the `event.once("Event.decrypted", ...)` approach today, but it's probably reasonable to publicise the promise for easier use in a context like this.